### ISSUE TITLE
`isolatedDeclarations` not giving related errors under `noEmit`

### ISSUE THREAD

**Issue by SuperchupuDev** [microsoft/TypeScript#59024]
### 🔎 Search Terms

"isolated declarations", "isolated declarations noEmit"

### 🕗 Version & Regression Information

- This changed in commit or PR e62418e8689738b93b627aac1116e02a48b0f9d1 (before that commit TS didn't recognize the option on tsconfig)

### 💻 Code

```ts
// must have a return type, should give error
function thingThatShouldErrorOnIsolatedDeclarations() {
  return 0;
}
```

### 🙁 Actual behavior

`tsc --noEmit --isolatedDeclarations` doesn't report any error, even though `tsc --isolatedDeclarations` does (vscode does too)

### 🙂 Expected behavior

`tsc --noEmit --isolatedDeclarations` reports the error, just like it does without the `--noEmit` flag

### Additional information about the issue

tsconfig used:
```json
{
  "compilerOptions": {
    "target": "es2022",
    "lib": ["esnext"],

    "module": "esnext",
    "moduleResolution": "bundler",

    "outDir": "dist",
    "esModuleInterop": true,
    "verbatimModuleSyntax": true,

    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true,
    "noEmitOnError": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "declaration": true,

    "strict": true
  }
}
```

**Comment by jakebailey** [comment]
Debugging, `handleNoEmitOptions` skips getting declaration diagnostics when `noEmit` is set. The editor is different and always grabs declaration errors.

Disabling the noEmit skip there fixes the issue, but then has weird knock-on effects. I haven't been able to solve. My test has been pushed to https://github.com/jakebailey/TypeScript/tree/fix-59024 (it doesn't work in a regular compiler test since we always emit).

@sheetalkamat Do you have any ideas as to how to fix this?

**Comment by sheetalkamat** [comment]
Will take a look.

**Comment by sheetalkamat** [comment]
This is not a regression. We never reported `declaration diagnostics` under `--noEmit`  through `tsc` 
If we are changing this it will be need to be not a patch.

**Comment by jakebailey** [comment]
Not a regression, no, but I feel like we need to do something here because we're offering up these diagnostics in the editor regardless of noEmit, but then you can get a passing result at the CLI or similar which seems really, really weird?

**Comment by sheetalkamat** [comment]
These diagnostics are offered when "emitting". Eg if you ran this code with "--noCheck" you will get that error. So its just where the error gets reported. It was modeled so that we are not doing extra pass .. so dts diagnostics are always reported when emitting dts file. (only excpetion is `noEmitOnError`)

**Comment by RyanCavanaugh** [comment]
Specifying `isolatedDeclarations: true` only makes sense if you want the errors (that's all the flag does, after all), so we should be evaluating them here. It's the same as specifying `verbatimModuleSyntax` and `noEmit` - the latter doesn't affect whether or not the former should do its checks
